File: C++/Data Structures/Binary Search Tree/binarySearchTree.cpp
//
// @author: Sayantan Paul
//

/*
 * Implementation of Binary Search Tree in C++
 */

#include <iostream>

class Node {
public:
    int data;
    Node* left;
    Node* right;

    // constructor
    Node(int info): data(info), left(NULL), right(NULL) {}

    // display
    void displayNode(Node* node) {
        std::cout << node->data << " " << std::endl;
    }
};


class bst {
private:
    Node* root;                 // root node of the Binary Search Tree
    int height = 0;             // height of the Binary Search Tree
    int num_nodes = 0;          // number of nodes in Binary Search Tree
public:
    bst();
    ~bst();

    Node* get_root();
    bool is_empty(); 

    Node* insert_node(Node*, Node*);       // insertion of node in the Binary Search Tree
    Node* search_node(Node*, Node*);       // searching of node in Binary Search Tree
    Node* delete_node(Node*, Node*);       // deleting node in Binary Search Tree

    Node* get_successor(Node*);             // returns node with highest value in left subtree 
    Node* get_predecessor(Node*);           // returns node with lowest value in right subtree

    int height_of_tree(Node*);  // calculate height of the Binary Search Tree
};

// constructor to initialize Binary Search Tree
bst::bst(): root(NULL) {}

// destructor destroys Binary Search Tree after execution of program is complete
bst::~bst() {
    destroy_tree(root);
    height = 0;
    num_nodes = 0;
}
// a recursive function to delete all nodes of the Binary Search Tree
void destroy_tree(Node* node) {
    if (node) {
        destroy_tree(node->left);
        destroy_tree(node->right);
        delete node;
    }
}

/**
 * @brief Check if Binary Search Tree is empty
 * 
 * @return true if empty
 * @return false if full
 */
bool bst::is_empty() {
    return root == NULL;
}


/**
 * @brief returns root node of Binary Search Tree
 * 
 * @return Node* 
 */
Node* bst::get_root() {
    return root;
}


/**
 * @brief Insertion Operation in Binary Search Tree
 *        Time Complexity: O(n), where n is total number of nodes
 * 
 * @param node 
 */
Node* bst::insert_node(Node* curr, Node* node) {
    if (curr == NULL) {
        curr = node;
        return curr;
    }

    if (curr->data < node->data) {
        curr->right = insert_node(curr->right, node);
    } else if (curr->data > node->data) {
        curr->left = insert_node(curr->left, node);
    }

    return curr;
}

/**
 * @brief Searching Operation in Binary Search Tree
 *        Time Complexity: O(h), where h is height of the tree
 * 
 * @param root 
 * @param node 
 * @return Node* 
 */
Node* bst::search_node(Node* root, Node* node) {
    if (root == NULL) return NULL;
    
    if (root->data == node->data) {
        return root;
    }

    if (root->data > node->data) {
        return search_node(root->left, node);
    } else {
        return search_node(root->right, node);
    }
}


/**
 * @brief Deletion Operation in Binary Search Tree
 *        Time Complexity: O(logn), where n is total number of nodes
 * 
 *        Deleting node consists of three cases:
 *          1. Deleted Node is leaf node
 *          2. Deleted Node has 1 child
 *          3. Deleted Node has 2 children
 * 
 * @param root 
 * @param node 
 * @return Node* return the node that is deleted
 */
Node* bst::delete_node(Node* root, Node* node) {
    if (root == NULL) return root;
    Node* tmp;

    // Finding the node by traversing left or right subtree recursively
    if (node->data > root->data) {
        delete_node(root->right, node);
    } else if (node->data < root->data) {
        delete_node(root->left, node);
    } else {
        // Now, we have found the node to be deleted
        // It can have no child or 1 child or 2 children

        // case 1: No child, i.e, this is a leaf node
        if (root->left == NULL && root->right == NULL) {
            std::cout << "Deleted Node is a leaf node" << std::endl;
            tmp = root;
            delete root;
            return tmp;
        }
        // case 2: 1 child, i.e, either left child or right child
        // Node has right child only
        if (root->left == NULL) {
            std::cout << "Deleted Node has a right-child" << std::endl;
            tmp = root->right;
            delete root;
            return tmp;
        } 
        // Node has left child only
        else if (root->right == NULL) {
            std::cout << "Deleted Node has a left-child" << std::endl;
            tmp = root->left;
            delete root;
            return tmp;
        }
        // case 2: 2 children, i.e, both left and right child
        std::cout << "Deleted Node has both left and right-child" << std::endl;
        tmp = get_successor(root->left);
        root->data = tmp->data;
        root->right = delete_node(root->right, tmp);
    }

    return root;
}


Node* bst::get_successor(Node* node) {
    if (node->left) {
        return get_successor(node->left);
    }

    return node;
}




int main() {

    return 0;
}

File: Python/Data_Structures/Binary Tree/binary_search_tree.py
"""
Implementation of Binary Search Tree in Python
"""


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def __str__(self):
        return str(self.data)

    def __repr__(self):
        return str(self.data)


class BinarySearchTree:
    # initialize the BST
    def __init__(self):
        self.root = None
        self.size = 0

    # insert a new node into the BST
    def insert(self, data):
        if not self.root:
            self.root = Node(data)
        else:
            self.insertNode(data, self.root)
        self.size += 1

    # insert a new node into the BST
    def insertNode(self, data, node):
        if data < node.data:
            if node.left:
                self.insertNode(data, node.left)
            else:
                node.left = Node(data)
        else:
            if node.right:
                self.insertNode(data, node.right)
            else:
                node.right = Node(data)

    # search for a node in the BST
    def search(self, data):
        if self.root:
            return self.searchNode(data, self.root)
        else:
            return False

    # search for a node in the BST
    def searchNode(self, data, node):
        if data == node.data:
            return True
        elif data < node.data and node.left:
            return self.searchNode(data, node.left)
        elif data > node.data and node.right:
            return self.searchNode(data, node.right)
        else:
            return False

    # print the BST as inorder traversal
    def inorder(self, node):
        if node:
            self.inorder(node.left)
            print(node.data, end=' ')
            self.inorder(node.right)

    # print the BST as preorder traversal
    def preorder(self, node):
        if node:
            print(node.data, end=' ')
            self.preorder(node.left)
            self.preorder(node.right)

    # print the BST as postorder traversal
    def postorder(self, node):
        if node:
            self.postorder(node.left)
            self.postorder(node.right)
            print(node.data, end=' ')

    # print the BST as levelorder traversal
    def printLevelorder(self):
        if self.root:
            self.printLevelorderNode(self.root)
        else:
            print("Tree is empty")

    # print the BST as levelorder traversal
    def printLevelorderNode(self, node):
        queue = [node]
        while queue:
            node = queue.pop(0)
            print(node.data)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

    # delete a node from the BST
    def delete(self, node: Node, data: int) -> Node:
        if node is None:
            return node

        # reach the position where the node to be deleted is present
        if data < node.data:
            node.left = self.delete(node.left, data)
        elif data > node.data:
            node.right = self.delete(node.right, data)
        else:
            # case 1: node has no child - a leaf node
            if node.left is None and node.right is None:
                return None

            # case 3: node has both left and right child
            elif node.left and node.right:
                # get the successor of the node to be deleted
                successor = self.get_successor(node)
                val = successor.data
                # delete the successor node
                self.delete(self.root, val)
                # set the value of the node as that of successor
                node.data = val

            else:
                # node has either left or right child
                child = node.left if node.left else node.right
                node = child
        return node

    def get_successor(self, node: Node) -> Node:
        return self.minimum_node(node.right)

    def minimum_node(self, node: Node) -> Node:
        while node.left:
            node = self.minimum_node(node.left)
        return node


if __name__ == "__main__":
    bst = BinarySearchTree()
    bst.insert(10)
    bst.insert(5)
    bst.insert(15)
    bst.insert(6)
    bst.insert(3)
    bst.insert(8)
    bst.insert(13)
    bst.insert(20)
    bst.insert(1)
    bst.insert(4)
    bst.insert(7)
    bst.insert(9)
    bst.insert(11)

    bst.inorder(bst.root)
    print()

    bst.delete(bst.root, 3)

    bst.inorder(bst.root)
    print()
